test/dataframe.jl

```diff
@@ -662,4 +662,104 @@ module TestDataFrame
+            (1.0:5.0, 2.0:6.0, 3.0:7.0),
+            ('a':'e', 'b':'f', 'c':'g'),
+            (["foo", "bar", "baz"], ["this", "that", "the other"], ["who", "what", "when"]),
+            ([:who, :what, :when], 'a':'c', ["foo", "bar", "baz"]),
+        ]
+
+        for (a, b, c) in column_sets
+            # One column
+            original = DataFrame(; a=a)
+
+            df = deepcopy(original)
+            expected = deepcopy(original)
+            permutecols!(df, [:a])
+            @test df == expected
+            permutecols!(df, [1])
+            @test df == expected
+
+            # Two columns
+            original = DataFrame(; a=a, b=b)
+
+            df = deepcopy(original)
+            expected = deepcopy(original)
+            permutecols!(df, [:a, :b])
+            @test df == expected
+            permutecols!(df, 1:2)
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; b=b, a=a)
+            permutecols!(df, [:b, :a])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [2, 1])
+            @test df == expected
+
+            # Three columns
+            original = DataFrame(; a=a, b=b, c=c)
+
+            df = deepcopy(original)
+            expected = deepcopy(original)
+            permutecols!(df, [:a, :b, :c])
+            @test df == expected
+            permutecols!(df, 1:3)
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; b=b, c=c, a=a)
+            permutecols!(df, [:b, :c, :a])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [2, 3, 1])
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; c=c, a=a, b=b)
+            permutecols!(df, [:c, :a, :b])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [3, 1, 2])
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; a=a, c=c, b=b)
+            permutecols!(df, [:a, :c, :b])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [1, 3, 2])
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; b=b, a=a, c=c)
+            permutecols!(df, [:b, :a, :c])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [2, 1, 3])
+            @test df == expected
+
+            df = deepcopy(original)
+            expected = DataFrame(; c=c, b=b, a=a)
+            permutecols!(df, [:c, :b, :a])
+            @test df == expected
+            df = deepcopy(original)
+            permutecols!(df, [3, 2, 1])
+            @test df == expected
+
+            # Invalid
+            df = DataFrame(; a=a, b=b, c=c)
+            @test_throws Exception permutecols!(df, [:a, :b])
```

Comment:
Should be `@test_throws ArgumentError ...` once the exception type is changed above.